dev/src/index.ts

```diff
@@ -1107,62 +1104,55 @@ export class Firestore {
    * for further attempts.
    *
    * @private
-   * @param attemptsRemaining The number of available attempts.
+   * @param methodName Name of the Veneer API endpoint that takes a request
+   * and GAX options.
    * @param requestTag A unique client-assigned identifier for this request.
    * @param func Method returning a Promise than can be retried.
-   * @param delayMs How long to wait before issuing a this retry. Defaults to
-   * zero.
    * @returns  - A Promise with the function's result if successful within
    * `attemptsRemaining`. Otherwise, returns the last rejected Promise.
    */
-  private _retry<T>(
-    attemptsRemaining: number,
+  private async _retry<T>(
+    methodName: string,
     requestTag: string,
-    func: () => Promise<T>,
-    delayMs = 0
+    func: () => Promise<T>
   ): Promise<T> {
-    const self = this;
+    const backoff = new ExponentialBackoff();
 
-    const currentDelay = delayMs;
-    const nextDelay = delayMs || 100;
+    let lastError: Error | undefined = undefined;
 
-    --attemptsRemaining;
-
-    return new Promise(resolve => {
-      setTimeout(resolve, currentDelay);
-    })
-      .then(func)
-      .then(result => {
-        self._lastSuccessfulRequest = new Date().getTime();
-        return result;
-      })
-      .catch(err => {
-        if (err.code !== undefined && err.code !== GRPC_UNAVAILABLE) {
-          logger(
-            'Firestore._retry',
-            requestTag,
-            'Request failed with unrecoverable error:',
-            err
-          );
-          return Promise.reject(err);
-        }
-        if (attemptsRemaining === 0) {
-          logger(
-            'Firestore._retry',
-            requestTag,
-            'Request failed with error:',
-            err
-          );
-          return Promise.reject(err);
-        }
+    for (let attempt = 0; attempt < MAX_REQUEST_RETRIES; ++attempt) {
+      if (lastError) {
         logger(
           'Firestore._retry',
           requestTag,
           'Retrying request that failed with error:',
-          err
+          lastError
         );
-        return self._retry(attemptsRemaining, requestTag, func, nextDelay);
-      });
+      }
+
+      try {
+        const result = await backoff.backoffAndWait().then(func);
```

Comment:
Mixing `await` and `then` in the same line is kind of confusing. Is it the same as

```ts
await backoff.backoffAndWait();
this._lastSuccessfulRequest = new Date().getTime();
return await func();
```

Reply:
It's not quite (the successful request should only be set after `await func()`). I cleaned it up.